### PR TITLE
feature: pouch copy in client and daemon side

### PR DESCRIPTION
**Pull request by knightXun**
### Ⅰ. Describe what this PR did
1. copy file from container
2. copy file to container

### CODE REVIEW

File: apis/types/container_copy.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package types
```
Comment (allencloud) [inline apis/types/container_copy.go]: We do not manually edit this file, but generate it by swagger, you could see the reference: https://github.com/alibaba/pouch/blob/master/apis/README.md

File: cli/cp.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"io"
+	"strings"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
```
Comment (allencloud) [inline cli/cp.go]: We should improve the import sequence by https://github.com/alibaba/pouch/blob/master/docs/contributions/code_styles.md#rule003---import-packages

File: daemon/mgr/container.go [hunk 3]
```diff
@@ -1217,6 +1230,312 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
...
+
+	stat, err = container.StatPath(resolvedPath, absPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// We need to rebase the archive entries if the last element of the
+	// resolved path was a symlink that was evaluated and is now different
+	// than the requested path. For example, if the given path was "/foo/bar/",
+	// but it resolved to "/var/lib/docker/containers/{id}/foo/baz/", we want
```
Comment (allencloud) [inline daemon/mgr/container.go]: I do not think the details here are proper.

File: apis/types/container_copy.go (hunk 1 above)
Comment (knightXun) [reply apis/types/container_copy.go]: ok

File: apis/server/container_bridge.go [hunk 4]
```diff
@@ -19,6 +19,9 @@ import (
 	"github.com/gorilla/mux"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+	"encoding/base64"
```
Comment (allencloud) [inline apis/server/container_bridge.go]: import sequence.

File: client/container_copy.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	if err != nil {
+		return types.ContainerPathStat{}, err
+	}
+	defer ensureReaderClosed(response)
+	return getContainerPathStatFromHeader(response.Header)
+}
+
+func ensureReaderClosed(response *Response) {
```
Comment (allencloud) [inline client/container_copy.go]: This function is redudant with https://github.com/alibaba/pouch/blob/master/client/utils.go#L18. 
Maybe we can re-use that.

File: pkg/httputils/http_utils.go [hunk 6]
```diff
@@ -5,6 +5,7 @@ import (
 	"strings"
 )
 
+
```
Comment (allencloud) [inline pkg/httputils/http_utils.go]: Please remove this blank line.

**Comment by allencloud** [comment]
I only add some initial comment. For this valuable part, I think we need to spend more time and patience to make this move on. @knightXun 
We will definitely help to make it work. Let us carry on. Thanks a lot.

**Comment by allencloud** [comment]
Any update on this? @knightXun

**Comment by knightXun** [comment]
now working on e2e test

**Comment by knightXun** [comment]
ping @allencloud

**Comment by allencloud** [comment]
I am afraid that we need to add more details in swagger.yml to define the API side. For adding new APIs, we should add definition in swagger.yml, like https://github.com/alibaba/pouch/blob/master/apis/README.md .

So please help to add that as well. I think we are almost there. Keep going. Thanks a lot. @knightXun

File: daemon/mgr/container_state.go [hunk 7]
```diff
@@ -1,10 +1,17 @@
...
 
 import (
+	"os"
+	"path/filepath"
 	"time"
 
 	"github.com/alibaba/pouch/apis/types"
+	"github.com/alibaba/pouch/pkg/errtypes"
 	"github.com/alibaba/pouch/pkg/utils"
+	"github.com/docker/docker/pkg/symlink"
```
Comment (allencloud) [inline daemon/mgr/container_state.go]: Please add a blank line before `github.com/docker/docker/pkg/symlink`.

**Comment by allencloud** [comment]
Any update on this? @knightXun 
Do you still work on this? 🐸 🦊

**Comment by wangforthinker** [comment]
Should we consider to make file mode(rwx) stay the same ? @knightXun @allencloud

File: cli/cp.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/alibaba/pouch/apis/types"
```
Comment (fuweid) [inline cli/cp.go]: split it into two groups. one for pouch and one for other sources.

File: apis/swagger.yml [hunk 9]
```diff
@@ -951,6 +951,92 @@ paths:
...
+          $ref: "#/responses/404ErrorResponse"
+        500:
+          $ref: "#/responses/500ErrorResponse"
+      parameters:
+        - name: "id"
+          in: "path"
+          required: true
+          description: "ID or name of the container"
+          type: "string"
+        - name: "SRCPATH"
```
Comment (fuweid) [inline apis/swagger.yml]: how about using `path` here?

File: apis/swagger.yml [hunk 10]
```diff
@@ -951,6 +951,92 @@ paths:
...
```
Comment (fuweid) [inline apis/swagger.yml]: how about using `path` here? it seems that there is no `destPath` in the description. in order to make it more clearly, I think `path` is good choice. BTW, there is no need to use all upper case for parameters .

File: apis/swagger.yml [hunk 11]
```diff
@@ -951,6 +951,92 @@ paths:
...
```
Comment (fuweid) [inline apis/swagger.yml]: how about using `path` here?

File: daemon/mgr/container.go [hunk 12]
```diff
@@ -1331,6 +1341,378 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
...
+	}
+
+	return mgr.containerStatPath(ctx, container, path)
+}
+
+// containerStatPath stats the filesystem resource at the specified path in this
+// container. Returns stat info about the resource.
+func (mgr *ContainerManager) containerStatPath(ctx context.Context, c *Container, path string) (stat *types.ContainerPathStat, err error) {
+	//There is a dead lock head
+	//c.Lock()
```
Comment (fuweid) [inline daemon/mgr/container.go]: ? if the code is useless, please remove it.

BTW, this action can stat path to `stoped/created` container, right?

File: daemon/mgr/container.go [hunk 13]
```diff
@@ -1331,6 +1341,378 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
...
+	defer mgr.Unmount(ctx, c)
+
+	attachedVolumes := map[string]struct{}{}
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		// release the container resources(network and containerio)
+		err = mgr.releaseContainerResources(c)
```
Comment (fuweid) [inline daemon/mgr/container.go]: I don't think it is good move for release. we cannot impact running container if pouch fails to do head action.

File: daemon/mgr/container.go [hunk 14]
```diff
@@ -1331,6 +1341,378 @@ func (mgr *ContainerManager) updateContainerResources(c *Container, resources ty
...
+func (mgr *ContainerManager) ContainerExtractToDir(ctx context.Context, name, path string, noOverwriteDirNonDir bool, content io.Reader) error {
+	container, err := mgr.container(name)
+	if err != nil {
+		return err
+	}
+
+	return mgr.containerExtractToDir(ctx, container, path, noOverwriteDirNonDir, content)
+}
+
+func setupRemappedRoot(config *config.Config) ([]idtools.IDMap, []idtools.IDMap, error) {
```
Comment (fuweid) [inline daemon/mgr/container.go]: if we don't need this action, I think we can remove it.

File: daemon/mgr/container.go [hunk 15]
```diff
@@ -165,6 +169,12 @@ type ContainerMgr interface {
 
 	// Commit commits an image from a container.
 	Commit(ctx context.Context, name string, options *types.ContainerCommitOptions) (*types.ContainerCommitResp, error)
+
+	ContainerStatPath(ctx context.Context, name string, path string) (stat *types.ContainerPathStat, err error)
```
Comment (fuweid) [inline daemon/mgr/container.go]: could you add the comment for functions?

Comment (knightXun) [reply daemon/mgr/container.go]: sure

**Comment by rudyfly** [comment]
The title should be `add` instead of `and`?

**Comment by knightXun** [comment]
I will split this pr into 4 or 5 parts.

**Comment by fuweid** [comment]
thanks @knightXun . And I close it and looking forward to the new PR :)

**Pull request closed without merging**